### ISSUE TITLE
[DoctrineBridge] Doctrine relationships with Abstract Classes - Application crashes

### ISSUE THREAD

**Issue by hlecorche** [symfony/symfony#40031]
**Symfony version(s) affected**: symfony/doctrine-bridge 4.4.19 - 5.2.2

**Description**  
When [Doctrine relationships with Abstract Classes and Interfaces](https://symfony.com/doc/current/doctrine/resolve_target_entity.html) are used, Symfony crashes with an 137 exit code (out of memory).

This problem is observed since the 23d692107c81fbd584019ba82cf0ea1fb2b72443 commit.

**How to reproduce**  
* Use [Doctrine relationships with Abstract Classes and Interfaces](https://symfony.com/doc/current/doctrine/resolve_target_entity.html)
* Use Doctrine (eg: `php bin/console doctrine:schema:validate`)

Tested with PHP 7.2.34, 7.3.26, 7.4.14 and 8.0.1

**Comment by jderusse** [comment]
Could you please provide a reproducer? 
I used the demo project and added the 2 class + interface + config as described in the documentation, but I don't have any error

**Comment by hlecorche** [comment]
@jderusse , I did some tests and I just understood that the problem is more complex.

Here is an example : https://github.com/hlecorche/symfony-issue-40031

This subscriber generates an infinite loop :
https://github.com/hlecorche/symfony-issue-40031/blob/master/src/SimulateBundleInVendor/EventSubscriber/MappingSubscriber.php

But I managed to solve my issue : Lines 43 and 49 avoid causing an infinite loop  :relaxed:

Thank you for your answer and sorry for the inconvenience...

**Comment by jderusse** [comment]
Thank you very much for your help, but it looks like the reproducer is not related.
I cloned the repo https://github.com/hlecorche/symfony-issue-40031, and used symfony 5.2.1 (version prior to the 23d6921 commit) and it also failed.
The issue in your repository is you are calling `em->getMetadataFor()` inside an event listener hooking on `loadClassMetadata` event. Your fix is right, you have to provide a mechanism that prevent infinite loop.

Regarding the original issue with "Doctrine relationships with Abstract Classes" do you know how to reproduce it?

**Comment by hlecorche** [comment]
In fact, the infinite loop exists since 2016.

But this fix works until 4.4.19 / 5.2.2 versions :

```php
public function loadClassMetadata(LoadClassMetadataEventArgs $eventArgs): void
    {
        $metadata = $eventArgs->getClassMetadata();

        if ($metadata->isMappedSuperclass) {
            return;
        }

        $className = $metadata->getName();

        if ($this->entityInVendorMappingTodo && is_subclass_of($className, UserInterfaceInVendor::class)) {
            //User class if loaded before EntityInVendor class
            $entityInVendorMetadata = $eventArgs->getEntityManager()->getMetadataFactory()->getMetadataFor(EntityInVendor::class);
            $this->mappEntityInVendor($entityInVendorMetadata, $metadata);

        } elseif ($this->entityInVendorMappingTodo && EntityInVendor::class === $className) {
            //EntityInVendor class if loaded before User class
            $userMetadata = $eventArgs->getEntityManager()->getMetadataFactory()->getMetadataFor(UserInterfaceInVendor::class);
            $this->mappEntityInVendor($metadata, $userMetadata);
        }
    }

    protected function mappEntityInVendor(ClassMetadataInfo $entityInVendorMetadata, ClassMetadataInfo $userMetadata): void
    {
        $this->entityInVendorMappingTodo = false;

        $entityInVendorMetadata->setAssociationOverride(
            'user',
            [
                'targetEntity' => $userMetadata->getName(),
                'fieldName' => 'user',
                'id' => true,
                'joinColumns' => [[
                    'name' => 'user_id',
                    'referencedColumnName' => $userMetadata->getSingleIdentifierColumnName(),
                    'onDelete' => 'CASCADE',
                ]],
            ]
        );
    }
```

With 4.4.19 / 5.2.2 versions, this fix no longer works and generates an infinite loop.

This new fix avoid the infinite loop with 4.4.19 / 5.2.2 versions (`$this->entityInVendorMappingTodo = false` has been moved) :

```php
    public function loadClassMetadata(LoadClassMetadataEventArgs $eventArgs): void
    {
        $metadata = $eventArgs->getClassMetadata();

        if ($metadata->isMappedSuperclass) {
            return;
        }

        $className = $metadata->getName();

        if ($this->entityInVendorMappingTodo && is_subclass_of($className, UserInterfaceInVendor::class)) {
            $this->entityInVendorMappingTodo = false; //avoid infinite loop - Fix https://github.com/symfony/symfony/issues/40031
            //User class if loaded before EntityInVendor class
            $entityInVendorMetadata = $eventArgs->getEntityManager()->getMetadataFactory()->getMetadataFor(EntityInVendor::class);
            $this->mappEntityInVendor($entityInVendorMetadata, $metadata);

        } elseif ($this->entityInVendorMappingTodo && EntityInVendor::class === $className) {
            $this->entityInVendorMappingTodo = false; //avoid infinite loop - Fix https://github.com/symfony/symfony/issues/40031
            //EntityInVendor class if loaded before User class
            $userMetadata = $eventArgs->getEntityManager()->getMetadataFactory()->getMetadataFor(UserInterfaceInVendor::class);
            $this->mappEntityInVendor($metadata, $userMetadata);
        }
    }

    protected function mappEntityInVendor(ClassMetadataInfo $entityInVendorMetadata, ClassMetadataInfo $userMetadata): void
    {
        $entityInVendorMetadata->setAssociationOverride(
            'user',
            [
                'targetEntity' => $userMetadata->getName(),
                'fieldName' => 'user',
                'id' => true,
                'joinColumns' => [[
                    'name' => 'user_id',
                    'referencedColumnName' => $userMetadata->getSingleIdentifierColumnName(),
                    'onDelete' => 'CASCADE',
                ]],
            ]
        );
    }
```


I failed to reproduce the infinite loop with symfony-demo. But I succeeded with a bundle (it's strange). You can try with the "bundle" branch of my repository (https://github.com/hlecorche/symfony-issue-40031/tree/bundle) (this reproducer uses the old fix)  :

You can run :
* `php vendor/phpunit/phpunit/phpunit` (`boostrap.php` file loads the database)
* or `php tests/Functional/App/bin/console doctrine:schema:validate`


With 5.2.1 version:
```bash
$ php tests/Functional/App/bin/console doctrine:schema:validate
--> No error
```

With 5.2.2 version:
```bash
$ php tests/Functional/App/bin/console doctrine:schema:validate
mmap() failed: [12] Cannot allocate memory
```


But is this really a problem? Because the new fix is more logical and works with all versions ...

**Comment by jderusse** [comment]
:( sorry, the first snippet you gave (the one with `$this->entityInVendorMappingTodo = false;` in the `mappEntityInVendor` method) does not work with 5.2.1 on my machine.

What I did:
- clone your repo
- `composer up`
- `../symfony/link .` (this link my local symfony project with your app)
- cd ../symfony && git checkout v5.2.1
- cd - && ./vendor/bin/simple-phpunit => falil

**Comment by hlecorche** [comment]
It's strange 😕  Because with GitHub Actions :
* The test works with SF 5.2.1
* The test fails with SF 5.2.2 (timeout -> infinite loop)

https://github.com/hlecorche/symfony-issue-40031/actions/runs/524019385

I have the same behavior on my machine.

Did you use the "bundle" branch (not master) of my repository ?

**Comment by jderusse** [comment]
Hey, indeed, I'm able to reproduce the issue with the the `bundle` branch.
The issue, is because the PR #39990 changed the priority of the listeners. By being lazy loaded, the EventSubscribers now have a lower priority than EventListeners. So your custom `MapingListener` listener has now a Higher priority than the builtin `ResolveTargetEntityListener` eventSubscriber.
This issue we be fixed by #39978, but this one is an new feature. 

I opened #40042 to restore the previous priority and fix your issue. Could you give a try

**Comment by hlecorche** [comment]
Thank you for your explanations.

Your fix works 👍 

Thank you very much for your patience and your work ☺️
